src/transformStateWithClones.js

```diff
@@ -6,8 +6,30 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let step = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      step = {};
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(step, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete step[key];
+      }
+    }
+
+    result.push({ ...step });
```

Comment:
Pushing the copy of the object to the result array should be done once after the `switch` block, not within each case. This adheres to the DRY principle and ensures that the copy is only pushed after all actions have been processed for the current iteration.